### PR TITLE
feat(schematics): ng-add schematics

### PR DESCRIPTION
**Pull request by sulco**
### Checklist

   - Issue number for this PR: #1676 (required)
   - Docs included?: yes (yes/no; required for all API/functional changes) 
   - Test units included?: no (yes/no; required) 
   - In a clean directory, `yarn install`, `yarn test` run successfully? yes (yes/no; required)

### Description

This adds `ng add` command that simplifies adding and setting up AngularFire to Angular apps.
It follows the practices described in the current setup instructions. I've also updated these to include the `ng-add` command (the "manual" instructions are still left available though).

### Code sample

To try this functionality out, in the project's folder run:

- `npm run build`
- `cd dist/packages-dist`
- `npm link`

Then create a sample Angular project wherever (ng new angularfire-dummy) and in that directory run

- `npm link @angular/fire`
- `ng add @angular/fire`

_This is my first contribution here but I've tried to be careful about preparing this - if you find anything unclear though, or in a need of change, please do bug me;)_

### CODE REVIEW

File: docs/install-and-setup.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 > Using Ionic and the Ionic CLI? Check out these [specific instructions](ionic/cli.md) for Ionic and their CLI.
 
+> Following instructions use `ng add` command that automates several steps for you. If you want to set up things manually refer to [these steps](install-and-setup-manual.md) instead.
```
Comment (Chan4077) [inline docs/install-and-setup.md]: Typos:

* "Following instructions" -> "The following instructions"
* "use `ng add` command" -> "use the `ng add` command"

File: docs/install-and-setup.md [hunk 2]
```diff
@@ -28,109 +30,47 @@ open http://localhost:4200
...
+By providing the `--all` flag you can also add modules for the individual @NgModules that your application needs:
 
  - `AngularFirestoreModule`
  - `AngularFireAuthModule`
  - `AngularFireDatabaseModule`
  - `AngularFireStorageModule`
  - `AngularFireMessagingModule` (Future release)
 
-#### Adding the Firebase Database and Auth Modules
+### 4. Edit Firebase config in environments variable
```
Comment (Chan4077) [inline docs/install-and-setup.md]: Probably simpler to just say "Edit Firebase configuration"?

File: docs/install-and-setup.md [hunk 3]
```diff
@@ -28,109 +30,47 @@ open http://localhost:4200
...
+export const environment = {
+  firebase: {
+    apiKey: '<your-key>',
+    authDomain: '<your-project-authdomain>',
+    databaseURL: '<your-database-URL>',
+    projectId: '<your-project-id>',
+    storageBucket: '<your-storage-bucket>',
+    messagingSenderId: '<your-messaging-sender-id>',
+  },
+  production: false,
```
Comment (Chan4077) [inline docs/install-and-setup.md]: Remove the leading comma here

File: docs/install-and-setup.md [hunk 4]
```diff
@@ -141,16 +81,14 @@ import { AngularFirestore } from 'angularfire2/firestore';
 @Component({
   selector: 'app-root',
   templateUrl: 'app.component.html',
-  styleUrls: ['app.component.css']
+  styleUrls: ['app.component.css'],
```
Comment (Chan4077) [inline docs/install-and-setup.md]: Once again, remove the leading comma here.

File: docs/install-and-setup.md [hunk 5]
```diff
@@ -162,7 +100,7 @@ import { Observable } from 'rxjs';
...
```
Comment (Chan4077) [inline docs/install-and-setup.md]: Remove leading comma here

File: schematics/src/ng-add/schema.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      "subtype": "filepath"
+    },
+    "firebaseApp": {
+      "type": "string",
+      "description": "A custom FirebaseApp name.",
+      "alias": "f"
+    },
+    "all": {
+      "type": "boolean",
+      "description": "Add all AngularFire modules (auth, db, storage).",
```
Comment (Chan4077) [inline schematics/src/ng-add/schema.json]: What about `AngularFireMessaging`?

File: schematics/src/ng-add/schema.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      "alias": "a",
+      "default": false
+    },
+    "path": {
+      "type": "string",
+      "format": "path",
+      "description": "The path to the project.",
```
Comment (Chan4077) [inline schematics/src/ng-add/schema.json]: You should be more specific here. Do you mean the project's _root_, or something else?

File: schematics/src/utils/libs-version.ts [hunk 8]
```diff
@@ -0,0 +1,2 @@
+export const angularfire2Version = '5.0.0-rc.11';
```
Comment (Chan4077) [inline schematics/src/utils/libs-version.ts]: There's a newer version of AngularFire2 as of 3rd September: `5.0.0-rc.12`

File: schematics/src/ng-add/schema.json (hunk 6 above)
Comment (Chan4077) [reply schematics/src/ng-add/schema.json]: Oops. Didn't see the line in the markdown file earlier.

**Comment by sulco** [comment]
Thanks @Chan4077, I've made the requested changes.

**Comment by sulco** [comment]
Hi @davideast! Yeah, I've seen that 😳Well, waiting is hard but actually leaving things undone is much worse, so full support to the plan. Let's wait for that landing and pick it up from there.

**Comment by sulco** [comment]
Ok @davideast, I've adjusted the code with the `angular/fire` changes (and also cleaned up the commits to more sensible 2)

**Comment by lundelius** [comment]
Hi Guys,
I just checked out the NG v7...

Would be lovely if you finish this PR for `ng add @angular/fire` !

**Comment by IOAyman** [comment]
Thanks everybody for the good work :)

Any updates or ETA for this, please?

/cc @davideast @Chan4077 @sulco

**Comment by sulco** [comment]
I can update this over the weekend so it doesn't break travis, and take a look if it still works after all these months, but other than that it's out of my hands. 
It just needs that review, I'll probably just bug the maintainers more for it.

**Comment by IOAyman** [comment]
@sulco That'd be great, thanks. Hopefully a collaborator will merge and release it soon.

@Chan4077 Sorry, my bad.

**Comment by Splaktar** [comment]
It looks like this needs a rebase now. I'd love to see this get in soon so that it could get tweaked a bit more (in a separate PR) for Angular version 8.

**Comment by sulco** [comment]
Hey, @Splaktar! Apologies for taking so long and thanks for the encouragement - I've just rebased it, updated relevant dependencies to the latest + updated the PR's description on how to try the functionality out.

Hopefully we're closing in on this one! :)

File: package.json [hunk 9]
```diff
@@ -30,11 +33,14 @@
...
   "homepage": "https://github.com/angular/angularfire2#readme",
   "dependencies": {
+    "@angular-devkit/core": "7.3.7",
+    "@angular-devkit/schematics": "7.3.7",
     "@angular/common": ">=6.0.0 <8",
     "@angular/compiler": ">=6.0.0 <8",
     "@angular/core": ">=6.0.0 <8",
     "@angular/platform-browser": ">=6.0.0 <8",
     "@angular/platform-browser-dynamic": ">=6.0.0 <8",
+    "bufferutil": "^3.0.3",
```
Comment (Chan4077) [inline package.json]: Just wondering, but what is `bufferutil` needed for here?

**Comment by sulco** [comment]
Thank you for pointing this one out @Chan4077 - tbh I don't remember, it's been so long since I've created this PR. I'll look that one up with once I gather the remaining feedback.

On this note. Are there any plans to have this reviewed and merged any time soon @Splaktar / @davideast? The new versions of - well, everything - are coming and going, and tbh I don't know if there is a point of just keep rebasing this PR every now and then ;)

**Comment by jamesdaniels** [comment]
@sulco Thanks for your work getting these base schematics rolling. We're centering work around #2046 which includes a deploy command to Firebase Hosting. Would you be able to look at whether the schematics you've built here would able be to ported over to that branch / directory style?

**Comment by sulco** [comment]
Looking at https://twitter.com/mgechev/status/1128164567459274752 I'm not sure if this work is relevant anymore - looks like you've got `ng-add` functionality covered in the meantime.

Kinda disappointing tbh. Anyways, please close/reject/... this PR if that's the case. Doesn't make sense to keep this one hanging.

**Comment by Splaktar** [comment]
Yes, I agree that not accepting or even reviewing this PR, letting it sit for 6+ months, and then re-doing the work in a separate branch/PR is NOT how open source should work. It does not provide a good experience to skilled contributors who spend a large amount of spare time to help the project. 

I understand that resources to maintain this repo are very limited, but that's even more of a reason that work like this should have been leveraged.

@jamesdaniels please try to find a way to ensure that something like this doesn't happen again 😞 Additionally, please set expectations properly for contributors in the docs here and when you are at conferences. If PRs, or certain types of PRs, won't be reviewed or accepted, then that needs to be communicated in order to avoid wasted time, frustration, etc.

**Comment by jamesdaniels** [comment]
I still believe there is value in this PR. As the ng-add and deploy commands as they exist on master don't overlap with the functionality of this PR. As mentioned on Apr 25th, I'd like to see if we can combine the efforts. Sorry for the lack of communication here, I too was caught off guard by the ng-deploy addition.

**Comment by jamesdaniels** [comment]
Also I don't believe communicating when I'm at conferences will be terribly useful, as that's pretty much my job 😝 think I'm on track to hit 30 this year. 

TBH I thought @davideast had this one and that was a failure in our communications. 

We are ramping up additional team members on the project and I simplified the release process with CI/CD (pending the Angular team allowing me to create the git triggers), which should help our agility.

**Comment by Splaktar** [comment]
> Also I don't believe communicating when I'm at conferences will be terribly useful, as that's pretty much my job

I think that there was some kind of misunderstanding here. It sounds like you are responding to a request for you to let contributors know when you will be at conferences. That wasn't at all what I was meaning.

Let me try rephrasing it...

When you are at conferences (as you mentioned, frequently), please talk about the expectations that contributors to `@angular/fire` should have. It would be nice to also cover contribution expectations in the docs. If PRs, or certain types of PRs, won't be reviewed or accepted, then it would be nice to communicate that in order to avoid wasted time, frustration, etc.

**Comment by Rofer11607** [comment]
@sulco ng add is no longer working, please see  #2285

**Comment by Splaktar** [comment]
@Rofer11607 I don't think that this issue has anything to do with the changes in this PR by @sulco, which have not been merged.

**Comment by jamesdaniels** [comment]
Closing in favor of #2836, where I brought over the feature initialization and environment config changes from this PR. Couple more things left to do but I'll include in 7.1, cutting a release candidate soon.

**Comment by jamesdaniels** [comment]
@sulco sorry it took so long to get this one merged in. Thanks for your contribution, you've been co-authored in the big schematics commit. 

`@schematics/angular` is now published, so I didn't have to bring in all that in. 

In #2836 I've modernized with the new API & really amped up the `ng add` & `firebase-tools` integration so you can select/create projects, sites, and download the env config.

**Comment by sulco** [comment]
Oh snap, thanks @jamesdaniels! I think that that's the longest "better late than never" I've ever experienced :D 
And thanks @Splaktar for caring about the OSS experience 🙌

**Pull request closed without merging**